### PR TITLE
Make the exceptions serializable

### PR DESCRIPTION
**Pull request by dakinggg**
In order for our custom exceptions to propagate correctly from a multiprocessing context, they need to be serializable. This PR does that, and adds a unit test to confirm that all our custom exceptions are serializable.

Based on https://github.com/mosaicml/llm-foundry/pull/1233

### CODE REVIEW

**Review by milocress — APPROVED** [review]
Good work! Thanks for helping fix this!

File: llmfoundry/utils/exceptions.py [hunk 1]
```diff
@@ -28,46 +28,91 @@
...
+
+class UserError(BaseContextualError):
     """Error thrown when an error is caused by user input."""
 
     error_attribution = 'UserError'
 
+    def __reduce__(self):
+        if self.__class__ == UserError:
+            raise NotImplementedError(
+                'BaseContextualError is a base class and cannot be pickled.',
```
Comment (irenedea) [inline llmfoundry/utils/exceptions.py:80]: Update error message to say `UserError`?

Comment (irenedea) [reply llmfoundry/utils/exceptions.py:80]: similar to subclasses below

File: tests/utils/test_exceptions.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        if issubclass(exception_class, foundry_exceptions.BaseContextualError)
+    ]
+    return exceptions
+
+
+@pytest.mark.parametrize(
+    'exception_class',
+    filter_exceptions(dir(foundry_exceptions)),
+)
+def test_exception_serialization(
```
Comment (irenedea) [inline tests/utils/test_exceptions.py:74]: this is nice!

File: llmfoundry/utils/exceptions.py [hunk 3]
```diff
@@ -28,46 +28,91 @@
...
+        self.serializable_attributes = []
 
-class UserError(ContextualError):
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+            self.serializable_attributes.append(key)
+
+        super().__init__(message)
+
+    def __reduce__(self):
```
Comment (irenedea) [inline llmfoundry/utils/exceptions.py:61]: could you add a comment describing how we need this for pickling/multiprocessing from places like dataset.map

**Review by irenedea — APPROVED** [review]
thank you!!

**Pull request merged**